x-pack/plugins/upgrade_assistant/server/routes/ml_snapshots.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { schema } from '@kbn/config-schema';
+import { API_BASE_PATH } from '../../common/constants';
+import { versionCheckHandlerWrapper } from '../lib/es_version_precheck';
+import { handleEsError } from '../shared_imports';
+import { RouteDependencies } from '../types';
+
+export function registerMlSnapshotRoutes({ router }: RouteDependencies) {
+  router.post(
+    {
+      path: `${API_BASE_PATH}/ml_snapshots`,
+      validate: {
+        body: schema.object({
+          snapshotId: schema.string(),
+          jobId: schema.string(),
+        }),
+      },
+    },
+    versionCheckHandlerWrapper(
+      async (
+        {
+          core: {
+            elasticsearch: { client },
+          },
+        },
+        request,
+        response
+      ) => {
+        try {
+          const { snapshotId, jobId } = request.body;
+
+          const {
+            body: upgradeSnapshotResponse,
+          } = await client.asCurrentUser.ml.updateModelSnapshot({
```

Comment:
So, there are two apis. 

Here you are referencing `updateModelSnapshot` which is for updating snapshot content (description, etc.)

What you want to use is [upgradeJobSnapshot](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html#_ml_upgradejobsnapshot)

Also, in the upgrade, you probably want to set the `timeout` to some sane value for Kibana (idk if there is a global timeout from kibana to ES, note, for VERY large models, this can take MINUTES to complete). 

`wait_for_completion` also makes this call synchronous (meaning the API won't get a response until its all finished).

Reply:
> Here you are referencing updateModelSnapshot which is for updating snapshot content (description, etc.)

Great catch! That was a mistake on my part. I had intended to use the upgrade API. Will address.

Reply:
> Also, in the upgrade, you probably want to set the timeout to some sane value for Kibana (idk if there is a global timeout from kibana to ES, note, for VERY large models, this can take MINUTES to complete).
> 
> wait_for_completion also makes this call synchronous (meaning the API won't get a response until its all finished).

@benwtrent I’m really glad you pointed this out. 

It looks like Kibana provides a default request timeout from ES to `3000ms` (set via `elasticsearch.requestTimeout` in `kibana.yml`)

We also support reindexing in UA. In that case, we do not wait for the request to complete, but instead use the tasks API to poll the status and update the UI accordingly. I’m wondering if a similar approach makes sense/would work here. I see in the [upgrade model snapshots API docs](https://www.elastic.co/guide/en/elasticsearch/reference/master/ml-upgrade-job-model-snapshot.html) the response returns `completed: false` if the task is still running. Is there a way to check the status afterward to verify it completed?

Reply:
@alisonelizabeth 100%. You can do something similar to the `_reindex` API. 

```
GET _tasks?nodes=Q1mqNskVT92KMdYkc9ZbbQ&actions=xpack/ml/job/snapshot/upgrade*&detailed
```
^ This is the API call to the `_tasks` API, using the node that is returned in the task response and the task action name. 

When the task is running it gives something like:

```
{
  "nodes" : {
    "Q1mqNskVT92KMdYkc9ZbbQ" : {
      "name" : "runTask-0",
      "transport_address" : "127.0.0.1:9300",
      "host" : "127.0.0.1",
      "ip" : "127.0.0.1:9300",
      "roles" : [
        "data",
        "data_cold",
        "data_content",
        "data_frozen",
        "data_hot",
        "data_warm",
        "ingest",
        "master",
        "ml",
        "remote_cluster_client",
        "transform"
      ],
      "attributes" : {
        "ml.machine_memory" : "68719476736",
        "xpack.installed" : "true",
        "testattr" : "test",
        "ml.max_open_jobs" : "512",
        "ml.max_jvm_size" : "1037959168"
      },
      "tasks" : {
        "Q1mqNskVT92KMdYkc9ZbbQ:38688" : {
          "node" : "Q1mqNskVT92KMdYkc9ZbbQ",
          "id" : 38688,
          "type" : "persistent",
          "action" : "xpack/ml/job/snapshot/upgrade[c]",
          "status" : {
            "state" : "STARTED"
          },
          "description" : "job-snapshot-upgrade-low_request_rate-1622116362",
          "start_time_in_millis" : 1622116801727,
          "running_time_in_nanos" : 1530756810,
          "cancellable" : true,
          "cancelled" : false,
          "parent_task_id" : "cluster:10",
          "headers" : { }
        }
      }
    }
  }
}
```
If there are multiple snapshot upgrades going on, you can filter again client side by the description `job-snapshot-upgrade-<job_id>-<snapshot_id>`

Once its done running, the task will simply disappear (sort of like _reindex).

Reply:
👍 thanks! I'll look into this approach. I may reach out with other questions as I start working on it 😄.